core/build.gradle

```diff
@@ -63,6 +63,7 @@ dependencies {
     api 'org.apache.calcite:calcite-linq4j:1.38.0'
     api project(':common')
     implementation "com.github.seancfoley:ipaddress:5.4.2"
+    implementation "com.google.zetasketch:zetasketch:0.1.0"
```

Comment:
is the HLL library Spark or OpenSearch using? Any other open source project use this library

Reply:
Spark use its own implementation of [HLL](https://github.com/apache/spark/blob/7e363747b15ef0dfbe4c56ad196886de8197b203/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala#L60). This dependencies is used by Scio,  Scala API for [Apache Beam](http://beam.incubator.apache.org/):https://github.com/spotify/scio/blob/3cf3f89de339ef7b9421d50cc738d28c62027165/scio-extra/src/main/scala/com/spotify/scio/extra/hll/zetasketch/HllPlus.scala#L23

Reply:
From the OpenSearch doc https://docs.opensearch.org/docs/latest/aggregations/metric/cardinality/#controlling-precision, the cardinality (count distinct) aggregation uses the HyperLogLog++ algorithm already. So as long as the count distinct is pushed down to DSL, it always is approximate. Please check which HLL library OpenSearch Core is leveraging. we'd better not to introduce a new dependency if core has already had. @xinyual

Reply:
Got it. I already change the code to use Opensearch core's implementation. Please check it.